calyx-py/calyx/queues.py

```diff
@@ -149,37 +142,311 @@ def __len__(self) -> int:
+    
+    We also take at initialization a `max_len` value to store the maximum possible
+    length of a queue.
+
+    When asked to push:
+    - If the PIEO is at length `max_len`, we raise an error.
+    - Otherwise, we insert the element into the PIEO such that the rank order stays increasing.
+
+    When asked to pop:
+    - If the length of `data` is 0, we raise an error .
+
+    - We can either pop based on value or based on eligibility.
+    - This implementation supports the most common eligibility predicate - whether an element is 'ripe'.
+
+    - If a value is passed in, we pop the first (lowest-rank) instance of that value.
+    - If no value is passed in but a time is,
+        we pop the first (lowest-rank) value that passes the predicate.
+    - Note that either a value or a bound must be passed in - both cannot be, nor can neither.
+
+    When asked to peek:
+    We do the same thing as `pop`, except:
+    - We peek into the PIEO instead of popping it - i.e. we don't remove any elements.
+
+    We compactly represent these similar operations through `query`, which takes in an additional
+    optional `remove` parameter (defaulted to False) to determine whether to pop or peek.
+    """
+    
+    def __init__(self, max_len: int):
+        """Initialize structure."""
+        self.max_len = max_len
+        self.data = []
+        self.insertion_count = 0
+
+    def __repr__(self):
+        return str(self.data)
+
+    def ripe(self, val, time):
+        """Check that a value is 'ripe' – i.e. its ready time has passed"""
+        return val[1] <= time
+    
+    def binsert(self, val, time, rank, l, r):
+        """Inserts element into list such that rank ordering is preserved
+        Uses variant of binary search algorithm
+        """
+        if l == r:
+            return self.data.insert(l, (val, time, rank))
+
+        mid = (l + r) // 2
+
+        if rank == self.data[mid][2]:
+            return self.data.insert(mid, (val, time, rank))
+
+        if rank > self.data[mid][2]:
+            return self.binsert(val, time, rank, mid+1, r)
+
+        if rank < self.data[mid][2]:
+            return self.binsert(val, time, rank, l, mid)
+        
+    def push(self, val, time=0, rank=0, insertion_count=None) -> None:
+        """Pushes to a PIEO.
+        Inserts element such that rank ordering is preserved
+        """
+        
+        #Breaks ties and maintains FIFO order (can pass either custom insertion order or use PIEO intenral one)
+        rank = (rank << 32) + (insertion_count or self.insertion_count)
+
+        if len(self.data) == self.max_len:
+            raise QueueError("Overflow")
+        
+        if len(self.data) == 0 or rank >= self.data[len(self.data)-1][2]:
+            self.data.append((val, time, rank))
+
+        elif rank <= self.data[0][2]:
+            self.data.insert(0, (val, time, rank))
+
+        else:
+            self.binsert(val, time, rank, 0, len(self.data))
```

Comment:
Kinda confused at first glance! Either explain this in comments, or swap to an equivalent library implementation!